corehq/apps/enterprise/enterprise.py

```diff
@@ -464,3 +569,18 @@ def rows_for_domain(self, domain_obj):
         num_errors = sum([result['direction_count'] for result in results if result['error']])
 
         return [(domain_obj.name, num_sent, num_received, num_errors), ]
+
+
+def _process_domains_in_parallel(process_func, domains):
+    """
+    Helper method to process domains in parallel using gevent pooling.
+    """
+    results = []
+
+    for chunk in chunked(domains, ES_CLIENT_CONNECTION_POOL_LIMIT):
+        pool = Pool(size=ES_CLIENT_CONNECTION_POOL_LIMIT)
+        chunk_results = pool.map(process_func, chunk)
+        pool.join()
+        results.extend(chunk_results)
+
+    return results
```

Comment:
Thanks for raising this concern Matt.

> If each of them tries to max out their elasticsearch connections, it seems like we're basically back to running operations in sequence, rather than in parallel?

I'm understanding this as a hypothetical right? Basically saying if we were to use this same technique on every tile, we would end up waiting for I/O elsewhere in the chain? But there is potentially still a world where we employ this technique on tile queries where it is especially useful? We don't have to necessarily max out the connection limit for this worker as I imagine there are still benefits to chunking this by as little as 2 or 3.

And just to reiterate, the larger concern I have is not with this performance change being moot, but it actually being harmful in that it could max out elasticsearch's search pool queue, as Jing noted, resulting in errors.

Reply:
Right, my concern was with adding needless complexity, if these performance improvements didn't turn out to be performance improvements. It sounds like you're concerned that they could lead to system stability issues, which is definitely more pressing. We definitely want to preserve the ability to selectively optimize specific queries, but I think we'd want to specifically call that out here with comments. Otherwise, I think there's a pretty high chance that this code gets copied as an example of how we write tile queries, and it leads to the situation I outlined, where every tile requests max capacity, resulting in no one really getting performance improvements